trie.c

```diff
@@ -110,59 +131,84 @@ bool st_find_longest_chain(st_cursor_t *cursor, st_trie_match_t *longest_match,
+            // Branch Node (with multiple children) if bit 14 is set
+            // st_debug(ST_DBG_SEQ_MATCH, "Branching Offset: %d; Code: %#04X", offset, code);
+            // code = TDATA(trie, ++offset);
             // Find child key that matches the search buffer at the current depth
             const uint8_t cur_key = st_cursor_get_triecode(cursor);
-            if (!cur_key || !find_branch_offset(trie, &offset, code, cur_key)) {
+            if (!cur_key || !find_branch_offset(trie, &offset, cur_key)) {
                 // Couldn't go deeper; return.
                 return longer_match_found;
             }
+            st_cursor_next(cursor);
         } else {
             // No high bits set, so this is a chain node
             // Travel down chain until we reach a zero byte, or we no longer match our buffer
-            code = TDATA(trie, ++offset);
-            do {
+            uint8_t code;
+            while ((code = TDATA(trie, offset++))) {
                 st_debug(ST_DBG_SEQ_MATCH, "Chaining Offset: %d; Code: %#04X\n", offset, code);
                 if (code != st_cursor_get_triecode(cursor))
                     return longer_match_found;
-            } while ((code = TDATA(trie, ++offset)) && st_cursor_next(cursor));
-            // After a chain, there should be a match or branch
-            ++offset;
-        }
-        // Traversed one (or more) buffer keys, check if we are at a match
-        code = TDATA(trie, offset);
-        // Match Node if bit 15 is set
-        if (code & TRIE_MATCH_BIT) {
-            st_debug(ST_DBG_SEQ_MATCH, "New Match found: (%d, %d) %d\n",
-                cursor->pos.index, cursor->pos.sub_index, cursor->pos.segment_len);
-            st_debug(ST_DBG_SEQ_MATCH, "Previous Match: (%d, %d) %d\n",
-                longest_match->seq_match_pos.index, longest_match->seq_match_pos.sub_index, longest_match->seq_match_pos.segment_len);
-            // record this if it is the longest match
-            if (st_cursor_longer_than(cursor, &longest_match->seq_match_pos)) {
-                longer_match_found = true;
-                longest_match->trie_match_index = offset;
-                longest_match->seq_match_pos = st_cursor_save(cursor);
-            }
-            // If bit 14 is also set, there is a child node after the completion string
-            if (code & TRIE_BRANCH_BIT) {
-                // move offset to next child node and continue walking the trie
-                offset += 4;
-                st_debug(ST_DBG_SEQ_MATCH, "  Looking for more: offset %d; code %d\n",
-                    offset, TDATA(trie, offset));
-            } else {
-                // No more matches; return
-                return longer_match_found;
+                st_cursor_next(cursor);
             }
+            // After a chain, there should be a match or branch
         }
-    } while (st_cursor_next(cursor));
-    return longer_match_found;
+    } while (true);
```

Comment:
The function no longer explicitly tests to see if it is at the end of the key buffer. Instead, we rely on the fact that all of the `st_cursor_get...``` functions will return 0 if the cursor is at the end, and 0 won't match to anything in the trie. Almost always in practice the trie match will end before the end of the buffer is reached, so the efficiency loss is minimal at worst, and possibly a win.

The advantage is that now the cursor can be preemptively advanced to the next key even if that key is not checked. This simplifies the code a bit.

The function is now written such that the cursor is always left at the position the next `get` should use and the `offset` is always left at the position that the next trie read should occur. The second one is important, because the next bit of code that uses the offset no longer always knows how far it needed to be advanced.